File: Linear Regression/implementation/cpp/utils/MathUtils.cpp
#include "./MathUtils.h"
#include <stdexcept>

//Calculate the mean of the x or y values in DataCollection
double MathUtils::calculateMean(const DataCollection* data, bool isX) {
    int size = data->getSize();
    if(size == 0) {
        throw std::invalid_argument("Cannot calculate mean value of an empty dataset");
    }

    double sum = 0;
    for(int i = 0; i < size; i++) {
        if(isX) {
            sum += data->getDataPoint(i).first; // Sum of x values 
        } else {
            sum += data->getDataPoint(i).second; // Sum of y values
        }
    }

    return sum / size;
}

//Calcualte the covariance between x and y in the DataCollection
double MathUtils::calculateCovariance(const DataCollection* data) {
    int size = data->getSize();
    if(size == 0) {
        throw std::invalid_argument("Cannot calculate covarience of an empty dataset");
    }

    double meanX = calculateMean(data, true);
    double meanY = calculateMean(data, false);

    double covarience = 0.0;
    for(int i = 0; i < size; ++i) {
        double x = data->getDataPoint(i).first;
        double y = data->getDataPoint(i).second;
        covarience += (x - meanX) * (y - meanY);
    }

    return covarience / (size - 1);
}

// Calculate the variance of the x values in DataCollection
double MathUtils::calculateVariance(const DataCollection* data) {
    int size = data->getSize();
    if(size == 0) {
        throw std::invalid_argument("Cannot calculate covarience of an empty dataset");
    }

    double meanX = calculateMean(data, true);

    double variance = 0.0;
    for(int i = 0; i < size; ++i) {
        double x = data->getDataPoint(i).first;
        variance += (x - meanX) * (x - meanX);
    }

    return variance / (size - 1);
}

File: Linear Regression/implementation/python/MathUtils.py
class MathUtils:
    @staticmethod
    def calculate_mean(values):
        return sum(values) / len(values)

    @staticmethod
    def calculate_covariance(x_values, y_values):
        mean_x = MathUtils.calculate_mean(x_values)
        mean_y = MathUtils.calculate_mean(y_values)
        
        covariance = sum((x - mean_x) * (y - mean_y) for x, y in zip(x_values, y_values))
        return covariance / (len(x_values) - 1)
    
    @staticmethod
    def calculate_variance(values):
        mean = MathUtils.calculate_mean(values)
        variance = sum((x - mean) ** 2 for x in values)
        return variance / (len(values) - 1)